lxd/api_cluster.go

```diff
@@ -2933,3 +2977,705 @@ func restoreClusterMember(d *Daemon, r *http.Request) response.Response {
+//           description: Status code
+//           example: 200
+//         metadata:
+//           $ref: "#/definitions/ClusterGroup"
+//   "403":
+//     $ref: "#/responses/Forbidden"
+//   "500":
+//     $ref: "#/responses/InternalServerError"
+func clusterGroupGet(d *Daemon, r *http.Request) response.Response {
+	name := mux.Vars(r)["name"]
+
+	clustered, err := cluster.Enabled(d.db)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if !clustered {
+		return response.BadRequest(fmt.Errorf("This server is not clustered"))
+	}
+
+	var group *db.ClusterGroup
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// Get the cluster group.
+		group, err = tx.GetClusterGroup(name)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		if errors.Is(err, db.ErrNoSuchObject) {
+			return response.NotFound(fmt.Errorf("Cluster group '%s' not found", name))
+		}
+
+		return response.SmartError(err)
+	}
+
+	apiGroup, err := group.ToAPI()
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	return response.SyncResponseETag(true, apiGroup, apiGroup.ClusterGroupPut)
+}
+
+// swagger:operation POST /1.0/cluster/groups/{name} cluster-groups cluster_group_post
+//
+// Rename the cluster group
+//
+// Renames an existing cluster group.
+//
+// ---
+// consumes:
+//   - application/json
+// produces:
+//   - application/json
+// parameters:
+//   - in: body
+//     name: name
+//     description: Cluster group rename request
+//     required: true
+//     schema:
+//       $ref: "#/definitions/ClusterGroupPost"
+// responses:
+//   "200":
+//     $ref: "#/responses/EmptySyncResponse"
+//   "400":
+//     $ref: "#/responses/BadRequest"
+//   "403":
+//     $ref: "#/responses/Forbidden"
+//   "500":
+//     $ref: "#/responses/InternalServerError"
+func clusterGroupPost(d *Daemon, r *http.Request) response.Response {
+	name := mux.Vars(r)["name"]
+	if name == "default" {
+		return response.Forbidden(errors.New(`The "default" group cannot be renamed`))
+	}
+
+	clustered, err := cluster.Enabled(d.db)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if !clustered {
+		return response.BadRequest(fmt.Errorf("This server is not clustered"))
+	}
+
+	req := api.ClusterGroupPost{}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		return response.BadRequest(err)
+	}
+
+	// Quick checks.
+	err = clusterGroupValidateName(name)
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// Check that the name isn't already in use.
+		_, err = tx.GetClusterGroup(req.Name)
+		if err == nil {
+			return fmt.Errorf("Name %q already in use", req.Name)
+		}
+
+		// Rename the cluster group.
+		err = tx.RenameClusterGroup(name, req.Name)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		if errors.Is(err, db.ErrNoSuchObject) {
+			return response.NotFound(fmt.Errorf("Cluster group '%s' not found", name))
```

Comment:
We're quoting strings using `%q` in errors nowadays for new stuff (and replacing them when we see them in older stuff).